tool/tctl/common/scoped_token_command.go

```diff
@@ -0,0 +1,358 @@
+	// client to verify the identity of the Auth Server.
+	localCAResponse, err := client.GetClusterCACert(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	caPins, err := tlsca.CalculatePins(localCAResponse.TLSCA)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Get list of auth servers. Used to print friendly signup message.
+	authServers, err := client.GetAuthServers()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if len(authServers) == 0 {
+		return trace.BadParameter("this cluster has no auth servers")
+	}
+
+	// Print signup message.
+	authServer := authServers[0].GetAddr()
+
+	pingResponse, err := client.Ping(ctx)
+	if err != nil {
+		slog.DebugContext(ctx, "unable to ping auth client", "error", err)
+	}
+
+	if err == nil && pingResponse.GetServerFeatures().Cloud {
+		proxies, err := client.GetProxies()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		if len(proxies) != 0 {
+			authServer = proxies[0].GetPublicAddr()
+		}
+	}
+
+	return nodeMessageTemplate.Execute(c.Stdout, map[string]any{
+		"token":        tokenName,
+		"scope":        tok.GetScope(),
+		"assign_scope": tok.GetSpec().GetAssignedScope(),
+		"roles":        strings.ToLower(roles.String()),
+		"minutes":      int(c.ttl.Minutes()),
+		"ca_pins":      caPins,
+		"auth_server":  authServer,
+	})
+}
+
+// Del is called to execute "scoped tokens del ..." command.
+func (c *ScopedTokensCommand) Del(ctx context.Context, client *authclient.Client) error {
+	if c.value == "" {
+		return trace.BadParameter("Need an argument: token")
+	}
+	if err := client.DeleteScopedToken(ctx, c.value); err != nil {
+		return trace.Wrap(err)
+	}
+	fmt.Fprintf(c.Stdout, "Token %s has been deleted\n", c.value)
+	return nil
+}
+
+// List is called to execute "tokens ls" command.
+func (c *ScopedTokensCommand) List(ctx context.Context, client *authclient.Client) error {
+	tokens, err := stream.Collect(clientutils.Resources(ctx, func(ctx context.Context, pageSize int, pageKey string) ([]*joiningv1.ScopedToken, string, error) {
+		res, err := client.ListScopedTokens(ctx, &joiningv1.ListScopedTokensRequest{
+			Limit:  uint32(pageSize),
+			Cursor: pageKey,
+		})
+		if err != nil {
+			return nil, "", trace.Wrap(err)
+		}
+
+		return res.GetTokens(), res.GetCursor(), nil
+	}))
+	if err != nil {
+		return trace.Wrap(err, "listing scoped tokens")
+	}
+
+	if len(tokens) == 0 && c.format == teleport.Text {
+		fmt.Fprintln(c.Stdout, "No active tokens found.")
+		return nil
+	}
+
+	// Sort by expire time.
+	slices.SortStableFunc(tokens, func(left, right *joiningv1.ScopedToken) int {
+		return left.GetMetadata().GetExpires().AsTime().Compare(right.GetMetadata().GetExpires().AsTime())
+	})
+
+	nameFunc := func(tok *joiningv1.ScopedToken) string {
+		if c.withSecrets {
+			return tok.GetMetadata().GetName()
+		}
+		scoped, err := token.NewScoped(tok)
+		if err != nil {
+			return tok.GetMetadata().GetName()
+		}
+
+		return scoped.GetSafeName()
```

Comment:
i'm not so sure we should print the full secret if token.NewScoped fails

if we worry about it failing, maybe pull out the GetSafeName logic so you can call it without token.NewScoped